### PR TITLE
Creates a default Ruleset Cursor, that fades out on idle.

### PR DESCRIPTION
**Pull request by Hanamuke**
RulesetContainer now has a default cursor, that fades out when idle for 2 seconds (delay that is inserted before the start of the beatmap), and pop back in when active. Previous behaviour was that MenuCursor was present throughout the game.

### CODE REVIEW

File: osu.Game/Rulesets/UI/Cursor/RulesetCursor.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+        protected override bool OnMouseMove(InputState state)
+        {
+            if (ActiveCursor.Position != state.Mouse.Position)
+            {
+                lastActiveTime = clock.CurrentTime;
+                if (State == Visibility.Hidden)
+                {
+                    Show();
+                    PopIn();
```
Comment (Aergwyn) [inline osu.Game/Rulesets/UI/Cursor/RulesetCursor.cs]: One of these should be enough. Same as below regarding hiding.

**Comment by AtomCrafty** [comment]
It think it would be useful to publicly expose the delay until the cursor disappears

**Comment by peppy** [comment]
Just to confirm: after I gave you advice on how to implement this you went a completely different (and incorrect) direction. Was there a reason for this or did I not make it clear enough how it should be done?

Notice that your method causes there to be *two* menu cursors on screen for periods when it switches between the "ruleset"'s and game's.

**Comment by Hanamuke** [comment]
You told me "it would be at a ruleset level" and "check if the position changed". Am I not at the ruleset level ? I did the same thing that GameplayCursor does. I feel like I followed your instructions, but I can change whatver you tell me to. I'm not sure what you call the game cursor, are there to identical cursor one on top of another ?

**Comment by peppy** [comment]
My proposal was to handle hiding of the cursor in the case the ruleset did not provide a cursors. I would see this being done in the base `RulesetContainer` or `Player`. 

While you could potentially do this the way you are, you would need to make your cursor implementation display nothing (and invert the logic). But even then, I don't believe it makes sense, as now if a ruleset returns null for the cursor it gives incorrect behaviour.

Also you could probably make use of #3415's idle tracking. It would probably go something like this (after using DI to get the `IdleTracker`):

image

**Comment by Hanamuke** [comment]
I've tried what you told me, but having a null cursor in RulesetContainer or setting ProvidingUserCursor to false in Player, only results in the default MenuCursor being provided. Besides, IdleTracker also tracks keyboard activity, and suffers from OnMouseMove being called even when the mouse doesn't move, so I can't really use it.  What do you think ?

Wit my patch, if a mania/taiko/catch player wants to use the mouse to click, they will have the cursor displayed. Wouldn't it be simpler to hide the cursor during play without bothering with idle time ?

**Comment by peppy** [comment]
The default menu cursor is what you want. Did you try exactly what i had in the screenshot? Was there an issue apart from it also considering keyboard? If not, just add a CursorIdle to idle tracker.

**Comment by Hanamuke** [comment]
Oh, i didn't understand that setting ProvidingUserCursor to true when cursor is null wouldn't crash but would instead display nothing. I'll implement your solution, however i'm not sure that I understand DI correctly. IdleTracker is a member of OsuGame, which instantiate PlayerLoader, which instantiate Players, does that mean that PlayerLoader needs to have the idle tracker as a member variable as well ?

**Comment by bdach** [comment]
Cursor hiding has been implemented since this PR has been opened (albeit in a slightly different manner), and the files that have been changed in this PR no longer even exist. Due to that, I'm going to close this PR as abandoned/superseded.

**Pull request closed without merging**